ntplib.py

```diff
@@ -297,37 +300,62 @@ def request(self, host, version=2, port="ntp", timeout=5, address_family=socket.
         NTPStats object
         """
 
-        # lookup server address
-        addrinfo = socket.getaddrinfo(host, port, address_family)[0]
-        family, sockaddr = addrinfo[0], addrinfo[4]
-
-        # create the socket
-        sock = socket.socket(family, socket.SOCK_DGRAM)
-
+        dest_timestamp = None
         try:
-            sock.settimeout(timeout)
-
-            # create the request packet - mode 3 is client
-            query_packet = NTPPacket(
-                mode=3,
-                version=version,
-                tx_timestamp=system_to_ntp_time(time.time())
-            )
-
-            # send the request
-            sock.sendto(query_packet.to_data(), sockaddr)
-
-            # wait for the response - check the source address
-            src_addr = (None,)
-            while src_addr[0] != sockaddr[0]:
-                response_packet, src_addr = sock.recvfrom(256)
-
-            # build the destination timestamp
-            dest_timestamp = system_to_ntp_time(time.time())
-        except socket.timeout:
-            raise NTPException("No response received from %s." % host)
-        finally:
-            sock.close()
+            # lookup server address - request only datagram (i.e. UDP) addresses
+            addrinfo_a = socket.getaddrinfo(host, port, address_family, socket.SOCK_DGRAM)
+        except socket.gaierror:
+            # can happen for a v6 only resolvable hostname on a v4 only host (or a v4 only hostname on a v6 only host)
+            # getaddrinfo() tries to be clever; which is damn annoying!
+            raise NTPException("No resolved address(es) for %s." % host)
+
+        # we randomize the addresses in order to spread the load
+        # if a specific address does not work, no worries, we just continue to the next
+        # for sanity reasons (i.e. limiting random sample cpu usage); we limit the tries to _MAX_HOSTS_TO_TRY addresses
+        for addrinfo in random.sample(addrinfo_a, k=min(NTPClient._MAX_HOSTS_TO_TRY, len(addrinfo_a))):
+            if addrinfo[1] != socket.SOCK_DGRAM or addrinfo[2] != socket.IPPROTO_UDP:
+                # should not happen - no worries, try the next one
+                continue
+
+            family, sockaddr = addrinfo[0], addrinfo[4]
+
+            # create the socket for UDP datagram
+            try:
+                sock = socket.socket(family, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
+            except OSError as e:
```

Comment:
Since it should not happen, I'd rather let the error propagate, to make debugging easier if it does happen.

Reply:
I think I have the same answer as above; however, with all this, it's your code. Feel free to delete.

That said, if you are on a machine with v6 not complied in to the kernel, I think this case can happen - and hence you want to `continue` and use the next returned address.  (I can't test this as I only have v6 enabled kernels). I'm mainly going after the randomization logic which then dictated the looping (once you realize the v4/v6 issue). The use of `SOCK_DGRAM` in the `socket.getaddrinfo()` call propagates downwards into the returned values.